### PR TITLE
Convert federation client to async/await.

### PR DESCRIPTION
**Pull request by clokep**
This converts the federation client and helper code to async/await, this code is really 3 layers:

* `federation.federation_client`
* `federation.transport.client`
* `http.matrixfederationclient`

All three are done as part of this (as well as a couple remaining methods in the `federation.sender`).

Most of this was straightforward, although the `groups_local` code was a bit tricky (and the tests took a bit to pass).

### CODE REVIEW

File: synapse/http/matrixfederationclient.py [hunk 1]
```diff
@@ -739,8 +739,7 @@ async def post_json(
         )
         return body
 
-    @defer.inlineCallbacks
-    def get_json(
+    async def get_json(
```
Comment (richvdh) [inline synapse/http/matrixfederationclient.py:739]: various things in TransportLayerClient that call this (`backfill`, `get_event`) have docstrings that claim to return Deferreds, which will no longer be true. It'd be nice to clean them up.

Comment (clokep) [reply synapse/http/matrixfederationclient.py:739]: Hmm...I thought I caught all those! I'll check again.

**Comment by clokep** [comment]
> (I found this slightly confusing to review: wouldn't mind smaller PRs next time?)

Sorry about that! This one got a little out of hand. 😢

File: synapse/http/matrixfederationclient.py (hunk 1 above)
Comment (clokep) [reply synapse/http/matrixfederationclient.py:739]: I found a few more of these, good call! 👍 Pushed an additional fix.

**Pull request merged**